File: src/main/java/com/aiproxyoauth/server/RequestValidator.java
package com.aiproxyoauth.server;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import io.javalin.http.Context;

public final class RequestValidator {

    private RequestValidator() {}

    public static JsonNode parseJsonObject(Context ctx, String body) throws JsonProcessingException {
        JsonNode parsed = JsonHelper.MAPPER.readTree(body);
        if (parsed == null || !parsed.isObject()) {
            JsonHelper.toErrorResponse(ctx, "Request body must be a JSON object.");
            return null;
        }
        return parsed;
    }

    public static boolean rejectMalformedJson(Context ctx, JsonProcessingException exception) {
        JsonHelper.toErrorResponse(ctx, "Malformed JSON request body.", 400, "invalid_request_error");
        return true;
    }
}

File: src/main/kotlin/de/moritzf/proxy/server/RequestValidator.kt
package de.moritzf.proxy.server
import de.moritzf.proxy.logging.RequestLogger
import io.ktor.server.request.receiveText
import kotlinx.serialization.json.JsonObject
object RequestValidator {
    suspend fun parseLoggedJsonObject(ctx: ProxyCall, requestLogger: RequestLogger, requestId: String): JsonObject? {
        val body = ctx.call.receiveText()
        requestLogger.logInbound(requestId, ctx, body)
        return try {
            parseJsonObject(ctx, body)
        } catch (_: Exception) {
            rejectMalformedJson(ctx)
            null
        }
    }
    suspend fun parseJsonObject(ctx: ProxyCall, body: String): JsonObject? {
        val parsed = JsonHelper.parseToJsonElement(body)
        if (parsed !is JsonObject) {
            JsonHelper.toErrorResponse(ctx, "Request body must be a JSON object.")
            return null
        }
        return parsed
    }
    suspend fun rejectMalformedJson(ctx: ProxyCall) {
        JsonHelper.toErrorResponse(ctx, "Malformed JSON request body.", 400, "invalid_request_error")
    }
}
